src/main/webapp/app/grading-system/grading-system.component.ts

```diff
@@ -118,12 +120,95 @@ export class GradingSystemComponent implements OnInit {
+                this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.nonUniqueGradeNames');
+                return false;
+            }
+            // check if the first passing grade is set if the grading scale is of type GRADE
+            if (this.firstPassingGrade === undefined || this.firstPassingGrade === '') {
+                this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.unsetFirstPassingGrade');
+                return false;
+            }
+        }
+        // copy the grade steps in a separate array, so they don't get dynamically updated when sorting
+        let sortedGradeSteps: GradeStep[] = [];
+        this.gradingScale.gradeSteps.forEach((gradeStep) => sortedGradeSteps.push(Object.assign({}, gradeStep)));
+        sortedGradeSteps = this.gradingSystemService.sortGradeSteps(sortedGradeSteps);
+        if (!this.isGradeType()) {
+            // check if when the grade type is BONUS the bonus points are at least 0
+            for (const gradeStep of sortedGradeSteps) {
+                if (isNaN(Number(gradeStep.gradeName)) || Number(gradeStep.gradeName) < 0) {
+                    this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.invalidBonusPoints');
+                    return false;
+                }
+            }
+            // check if when the grade type is BONUS the bonus points have strictly ascending values
+            if (
+                !sortedGradeSteps
+                    .map((gradeStep) => Number(gradeStep.gradeName))
+                    .every((bonusPoints, index, bonusPointsArray) => index === 0 || bonusPoints > bonusPointsArray[index - 1])
+            ) {
+                this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.nonStrictlyIncreasingBonusPoints');
+                return false;
+            }
+        }
+
+        // check if grade steps have valid adjacency
+        for (let i = 0; i < sortedGradeSteps.length - 1; i++) {
+            if (sortedGradeSteps[i].upperBoundPercentage !== sortedGradeSteps[i + 1].lowerBoundPercentage) {
+                this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.invalidAdjacency');
+                return false;
+            }
+        }
+        // check if first and last grade step are valid
+        if (sortedGradeSteps[0].lowerBoundPercentage !== 0 || sortedGradeSteps[sortedGradeSteps.length - 1].upperBoundPercentage !== 100) {
+            this.invalidGradeStepsMessage = this.translateService.instant('artemisApp.gradingSystem.error.invalidFirstAndLastStep');
+            return false;
+        }
+        this.invalidGradeStepsMessage = undefined;
+        return true;
+    }
+
```

Comment:
I thought about that while implementing and what you're saying makes sense. However, I kept it as is since the validation rules are all called at once every time anyways, and we have quite a few functions already in the client, that tend to do similar things in just one function (usually either calculations or validation). Therefore I opted for a more detailed explanation to still keep it readable.